test/core/client_channel/lb_policy/round_robin_test.cc

```diff
@@ -0,0 +1,373 @@
+  for (int i = 0; i < 3; i++) {
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  auto picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+  EXPECT_EQ(3, subchannel_pool_.size());
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, OneChannelReadyToIdle) {
+  auto policy = MakeLbPolicy("round_robin");
+  auto status = ApplyUpdate(BuildUpdateArgs({
+                                "ipv4:127.0.0.1:441",
+                                "ipv4:127.0.0.1:442",
+                                "ipv4:127.0.0.1:443",
+                            }),
+                            policy.get());
+  ASSERT_TRUE(status.ok()) << status;
+  for (int i = 0; i < 3; i++) {
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  auto subchannel = FindSubchannel("ipv4:127.0.0.1:441");
+  ASSERT_NE(subchannel, nullptr);
+  subchannel->SetConnectivityState(GRPC_CHANNEL_READY, absl::OkStatus());
+  ExpectState(GRPC_CHANNEL_READY);
+  subchannel->SetConnectivityState(GRPC_CHANNEL_IDLE, absl::OkStatus());
+  ExpectReresolutionRequest();
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, AllTransientFailure) {
+  std::array<std::string, 3> uris = {
+      "ipv4:127.0.0.1:441",
+      "ipv4:127.0.0.1:442",
+      "ipv4:127.0.0.1:443",
+  };
+  auto policy = MakeLbPolicy("round_robin");
+  auto status = ApplyUpdate(BuildUpdateArgs(uris), policy.get());
+  ASSERT_TRUE(status.ok()) << status;
+  for (int i = 0; i < 3; i++) {
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  for (size_t i = 0; i < uris.size() - i; ++i) {
+    auto subchannel = FindSubchannel(uris[i]);
+    ASSERT_NE(subchannel, nullptr);
+    subchannel->SetConnectivityState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+                                     absl::OkStatus());
+    ExpectReresolutionRequest();
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  FindSubchannel(uris[2])->SetConnectivityState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+                                                absl::OkStatus());
+  ExpectReresolutionRequest();
+  ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, NoChannels) {
+  auto policy = MakeLbPolicy("round_robin");
+  LoadBalancingPolicy::UpdateArgs update_args;
+  update_args.resolution_note = "This is a test";
+  update_args.addresses.emplace();
+  absl::Status status = ApplyUpdate(std::move(update_args), policy.get());
+  EXPECT_TRUE(absl::IsUnavailable(status));
+  ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+              absl::UnavailableError("empty address list: This is a test"));
+}
+
+TEST_F(RoundRobinTest, AddressListChange) {
+  std::array<std::string, 3> uris = {
+      "ipv4:127.0.0.1:441",
+      "ipv4:127.0.0.1:442",
+      "ipv4:127.0.0.1:443",
+  };
+  auto policy = MakeLbPolicy("round_robin");
+
+  // One channel that is ready. Gets picked all the time.
```

Comment:
Please change "channel" to "subchannel".

Same thing throughout.  There is nothing in this test that actually uses a channel, everything here is subchannels.

Reply:
Done